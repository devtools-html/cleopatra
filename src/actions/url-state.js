/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';

export function urlStateHasChanged(): Action {
  return {
    type: 'URL_STATE_HAS_CHANGED',
  };
}
