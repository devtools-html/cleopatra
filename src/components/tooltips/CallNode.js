/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
// import colors from 'photon-colors';
// import {
//   withChartViewport,
//   type WithChartViewport,
// } from '../shared/chart/Viewport';
// import NodeIcon from '../shared/NodeIcon';
// import ChartCanvas from '../shared/chart/Canvas';
// import TextMeasurement from '../../utils/text-measurement';
// import { getStackType } from '../../profile-logic/transforms';
//
// import type { Thread } from '../../types/profile';
// import type { CssPixels } from '../../types/units';
// import type {
//   FlameGraphTiming,
//   FlameGraphDepth,
//   IndexIntoFlameGraphTiming,
// } from '../../profile-logic/flame-graph';
//
// import type {
//   CallNodeInfo,
//   IndexIntoCallNodeTable,
//   StackType,
// } from '../../types/profile-derived';
// import type { CallTree } from '../../profile-logic/call-tree';

// export type Props = {|
//   +thread: Thread,
//   +callNodeIndex: IndexIntoCallNodeTable,
//   +maxStackDepth: number,
//   +flameGraphTiming: FlameGraphTiming,
//   +callNodeInfo: CallNodeInfo,
//   +callTree: CallTree,
//   +stackFrameHeight: CssPixels,
//   +selectedCallNodeIndex: IndexIntoCallNodeTable | null,
//   +onSelectionChange: (IndexIntoCallNodeTable | null) => void,
//   +disableTooltips: boolean,
//   +scrollToSelectionGeneration: number,
// |};

/* eslint-disable react/prefer-stateless-function */
export class TooltipCallNode extends React.PureComponent<{||}> {
  render() {
    return null;
    // const {
    //   thread,
    //   callNodeIndex,
    //   callTree,
    //   callNodeInfo: { callNodeTable },
    //   disableTooltips,
    // } = this.props;
    //
    // if (disableTooltips) {
    //   return null;
    // }
    //
    // // TODO - Reviewer, remind me of this.
    // // const stackTiming = flameGraphTiming[depth];
    // // const duration =
    // //   stackTiming.end[flameGraphTimingIndex] -
    // //   stackTiming.start[flameGraphTimingIndex];
    // const duration = 0;
    //
    // const funcIndex = callNodeTable.func[callNodeIndex];
    // const funcName = thread.stringTable.getString(
    //   thread.funcTable.name[funcIndex]
    // );
    //
    // const stackType = getStackType(thread, funcIndex);
    // const background = getBackgroundColor(
    //   stackType,
    //   stackTiming.selfTimeRelative[flameGraphTimingIndex]
    // );
    //
    // let stackTypeLabel;
    // switch (stackType) {
    //   case 'native':
    //     stackTypeLabel = 'Native';
    //     break;
    //   case 'js':
    //     stackTypeLabel = 'JavaScript';
    //     break;
    //   case 'unsymbolicated':
    //     stackTypeLabel = 'Unsymbolicated Native';
    //     break;
    //   default:
    //     throw new Error(`Unknown stack type case "${stackType}".`);
    // }
    //
    // const displayData = callTree.getDisplayData(callNodeIndex);
    //
    // return (
    //   <div className="flameGraphCanvasTooltip">
    //     <div className="tooltipHeader">
    //       <div className="tooltipTiming">{(100 * duration).toFixed(2)}%</div>
    //       <div className="tooltipTitle">{funcName}</div>
    //       <div className="tooltipIcon">
    //         {displayData.icon ? <NodeIcon displayData={displayData} /> : null}
    //       </div>
    //       <div className="tooltipLib">{displayData.lib}</div>
    //     </div>
    //     <div className="tooltipDetails">
    //       <div className="tooltipLabel">Stack Type:</div>
    //       <div>
    //         <div
    //           className="tooltipSwatch"
    //           style={{ backgroundColor: background }}
    //         />
    //         {stackTypeLabel}
    //       </div>
    //       <div className="tooltipLabel">Running Time:</div>
    //       <div>{displayData.totalTimeWithUnit}</div>
    //       <div className="tooltipLabel">Self Time:</div>
    //       <div>{displayData.selfTimeWithUnit}</div>
    //     </div>
    //   </div>
    // );
  }
}
