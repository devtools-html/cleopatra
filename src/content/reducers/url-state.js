// @flow
import { combineReducers } from 'redux';
import { defaultThreadOrder } from '../profile-data';
import { createSelector } from 'reselect';
import { urlFromState } from '../url-handling';
import * as RangeFilters from '../range-filters';

import type { ThreadIndex } from '../../common/types/profile';
import type { StartEndRange } from '../../common/types/units';
import type { Action, CallTreeFiltersPerThread, CallTreeFilter, DataSource } from '../actions/types';
import type { State, URLState, Reducer } from './types';

function dataSource(state: DataSource = 'none', action: Action) {
  switch (action.type) {
    case 'WAITING_FOR_PROFILE_FROM_FILE':
      return 'from-file';
    case 'PROFILE_PUBLISHED':
      return 'public';
    default:
      return state;
  }
}

function hash(state: string = '', action: Action) {
  switch (action.type) {
    case 'PROFILE_PUBLISHED':
      return action.hash;
    default:
      return state;
  }
}

function selectedTab(state: string = 'calltree', action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_TAB':
      return action.selectedTab;
    default:
      return state;
  }
}

function rangeFilters(state: StartEndRange[] = [], action: Action) {
  switch (action.type) {
    case 'ADD_RANGE_FILTER': {
      const { start, end } = action;
      return [...state, { start, end }];
    }
    case 'POP_RANGE_FILTERS':
      return state.slice(0, action.firstRemovedFilterIndex);
    default:
      return state;
  }
}

function selectedThread(state: ThreadIndex = 0, action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_THREAD':
      return action.selectedThread;
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_FILE': {
      // When loading in a brand new profile, select either the GeckoMain [tab] thread,
      // or the first thread in the thread order. For profiles from the Web, the
      // selectedThread has already been initialized from the URL and does not require
      // looking at the profile.
      const contentThreadId = action.profile.threads.findIndex(thread => {
        return thread.name === 'GeckoMain' && thread.processType === 'tab';
      });
      return contentThreadId !== -1 ? contentThreadId : defaultThreadOrder(action.profile.threads)[0];
    }
    default:
      return state;
  }
}

function callTreeSearchString(state: string = '', action: Action) {
  switch (action.type) {
    case 'CHANGE_CALL_TREE_SEARCH_STRING':
      return action.searchString;
    default:
      return state;
  }
}

function callTreeFilters(state: CallTreeFiltersPerThread = {}, action: Action) {
  switch (action.type) {
    case 'ADD_CALL_TREE_FILTER': {
      const { threadIndex, filter } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: [...oldFilters, filter],
      });
    }
    case 'POP_CALL_TREE_FILTERS': {
      const { threadIndex, firstRemovedFilterIndex } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: oldFilters.slice(0, firstRemovedFilterIndex),
      });
    }
    default:
      return state;
  }
}

function jsOnly(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_JS_ONLY':
      return action.jsOnly;
    default:
      return state;
  }
}

function invertCallstack(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
      return action.invertCallstack;
    default:
      return state;
  }
}

function hidePlatformDetails(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_HIDE_PLATFORM_DETAILS':
      return action.hidePlatformDetails;
    default:
      return state;
  }
}

const urlStateReducer: Reducer<URLState> = (regularUrlStateReducer => (state: URLState, action: Action): URLState => {
  switch (action.type) {
    case '@@urlenhancer/updateURLState':
      return action.urlState;
    default:
      return regularUrlStateReducer(state, action);
  }
})(combineReducers({
  dataSource, hash, selectedTab, rangeFilters, selectedThread,
  callTreeSearchString, callTreeFilters, jsOnly, invertCallstack,
  hidePlatformDetails,
}));
export default urlStateReducer;

const getURLState = (state: State): URLState => state.urlState;

export const getDataSource = (state: State) => getURLState(state).dataSource;
export const getHash = (state: State) => getURLState(state).hash;
export const getRangeFilters = (state: State) => getURLState(state).rangeFilters;
export const getJSOnly = (state: State) => getURLState(state).jsOnly;
export const getHidePlatformDetails = (state: State) => getURLState(state).hidePlatformDetails;
export const getInvertCallstack = (state: State) => getURLState(state).invertCallstack;
export const getSearchString = (state: State) => getURLState(state).callTreeSearchString;
export const getSelectedTab = (state: State) => getURLState(state).selectedTab;
export const getSelectedThreadIndex = (state: State) => getURLState(state).selectedThread;
export const getCallTreeFilters = (state: State, threadIndex: ThreadIndex): CallTreeFilter[] => {
  return getURLState(state).callTreeFilters[threadIndex] || [];
};

export const getURLPredictor = createSelector(
  getURLState,
  (oldURLState: URLState) => actionOrActionList => {
    const actionList = ('type' in actionOrActionList) ? [actionOrActionList] : actionOrActionList;
    const newURLState = actionList.reduce(urlStateReducer, oldURLState);
    return urlFromState(newURLState);
  }
);

export const getRangeFilterLabels = createSelector(
  getRangeFilters,
  RangeFilters.getRangeFilterLabels
);
