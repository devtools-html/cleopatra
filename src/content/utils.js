// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
