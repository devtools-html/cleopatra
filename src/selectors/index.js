/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import { selectedThreadSelectors as selectedThread } from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';

// Normally we don't need that type because flow can infer types, but on that
// object, it gives weird flow errors that say missing type annotation for X.
// Only adding a type like that seems to solve it.
type Selectors = {
  app: typeof app,
  profile: typeof profile,
  urlState: typeof urlState,
  icons: typeof icons,
  publish: typeof publish,
  zippedProfiles: typeof zippedProfiles,
  selectedThread: typeof selectedThread,
};

export default ({
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
}: Selectors);
