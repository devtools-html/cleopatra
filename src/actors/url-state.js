/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { getUrlState } from '../reducers/url-state';
import {
  urlStateHasChanged,
  updateUrlState,
  show404,
} from '../actions/url-state';
import { urlFromState, stateFromLocation } from '../url-handling';

import type { Dispatch } from '../types/store';
import type { State, UrlState } from '../types/reducers';

let previousUrlState = null;
export default function stateWatcher(state: State, dispatch: Dispatch) {
  function updateState() {
    if (window.history.state) {
      dispatch(updateUrlState(window.history.state));
    } else {
      try {
        const urlState = stateFromLocation(window.location);
        // replacing the existing location ensures we get the latest format for the URL
        updateLocation(urlState, true);
        dispatch(updateUrlState(urlState));
      } catch (e) {
        console.error(e);
        show404(window.location.pathname + window.location.search);
      }
    }
  }

  function initialSetup() {
    updateState();
    window.addEventListener('popstate', updateState);
  }

  function updateLocation(urlState: UrlState, replace: boolean = false) {
    const newUrl = urlFromState(urlState);
    const operation = replace
      ? window.history.replaceState.bind(window.history)
      : window.history.pushState.bind(window.history);
    if (newUrl !== window.location.pathname + window.location.search) {
      operation(urlState, document.title, newUrl);
    }
  }

  const newUrlState = getUrlState(state);
  if (!previousUrlState) {
    // first run
    previousUrlState = newUrlState;
    initialSetup();
  } else if (newUrlState !== previousUrlState) {
    previousUrlState = newUrlState;
    updateLocation(newUrlState);
    dispatch(urlStateHasChanged());
  }
}
