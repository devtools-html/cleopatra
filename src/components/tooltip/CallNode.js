/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
import * as React from 'react';

import { getStackType } from '../../profile-logic/transforms';
import { objectEntries } from '../../utils/flow';
import { formatNumberDependingOnInterval } from '../../utils/format-numbers';
import NodeIcon from '../shared/NodeIcon';
import { getFriendlyStackTypeName } from '../../profile-logic/profile-data';

import type { CallTree } from '../../profile-logic/call-tree';
import type { Thread, CategoryList } from '../../types/profile';
import type {
  IndexIntoCallNodeTable,
  CallNodeDisplayData,
  CallNodeInfo,
} from '../../types/profile-derived';
import type { TimingsForPath } from '../../profile-logic/profile-data';
import type { Milliseconds } from '../../types/units';

import './CallNode.css';

const GRAPH_WIDTH = 150;
const GRAPH_HEIGHT = 10;

type Props = {|
  thread: Thread,
  callNodeIndex: IndexIntoCallNodeTable,
  callNodeInfo: CallNodeInfo,
  categories: CategoryList,
  interval: Milliseconds,
  // Since this tooltip can be used in different context, provide some kind of duration
  // label, e.g. "100ms" or "33%".
  durationText: string,
  callTree?: CallTree,
  timings?: TimingsForPath,
|};

/**
 * This class collects the tooltip rendering for anything that cares about call nodes.
 * This includes the Flame Graph and Stack Chart.
 */
export class TooltipCallNode extends React.PureComponent<Props> {
  _renderTimings(
    maybeTimings: ?TimingsForPath,
    maybeDisplayData: ?CallNodeDisplayData
  ) {
    if (!maybeTimings || !maybeDisplayData) {
      return null;
    }
    const { totalTime, selfTime } = maybeTimings.forPath;
    const displayData = maybeDisplayData;
    if (!totalTime.breakdownByImplementation) {
      return null;
    }

    const sortedTotalBreakdownByImplementation = objectEntries(
      totalTime.breakdownByImplementation
    ).sort((a, b) => b[1] - a[1]);
    const { interval } = this.props;
    const isIntegerInterval = Number.isInteger(interval);

    return (
      <div className="tooltipCallNodeImplementation">
        {/* grid row -------------------------------------------------- */}
        <div />
        <div className="tooltipCallNodeImplementationHeader" />
        <div className="tooltipCallNodeImplementationHeader">
          <span className="tooltipCallNodeImplementationHeaderSwatchRunning" />
          Running
        </div>
        <div className="tooltipCallNodeImplementationHeader">
          <span className="tooltipCallNodeImplementationHeaderSwatchSelf" />
          Self
        </div>
        {/* grid row -------------------------------------------------- */}
        <div className="tooltipLabel">Overall</div>
        <div className="tooltipCallNodeImplementationGraph">
          <div
            className="tooltipCallNodeImplementationGraphRunning"
            style={{
              width: GRAPH_WIDTH,
            }}
          />
          <div
            className="tooltipCallNodeImplementationGraphSelf"
            style={{
              width: (GRAPH_WIDTH * selfTime.value) / totalTime.value,
            }}
          />
        </div>
        <div>{displayData.totalTimeWithUnit}</div>
        <div>{displayData.selfTimeWithUnit}</div>
        {/* grid row -------------------------------------------------- */}
        {sortedTotalBreakdownByImplementation.map(
          ([implementation, time], index) => {
            let selfTimeValue = 0;
            if (selfTime.breakdownByImplementation) {
              selfTimeValue =
                selfTime.breakdownByImplementation[implementation] || 0;
            }

            return (
              <React.Fragment key={index}>
                <div className="tooltipCallNodeImplementationName tooltipLabel">
                  {getFriendlyStackTypeName(implementation)}
                </div>
                <div className="tooltipCallNodeImplementationGraph">
                  <div
                    className="tooltipCallNodeImplementationGraphRunning"
                    style={{
                      width: (GRAPH_WIDTH * time) / totalTime.value,
                    }}
                  />
                  <div
                    className="tooltipCallNodeImplementationGraphSelf"
                    style={{
                      width: (GRAPH_WIDTH * selfTimeValue) / totalTime.value,
                    }}
                  />
                </div>
                <div className="tooltipCallNodeImplementationTiming">
                  {formatNumberDependingOnInterval(isIntegerInterval, time)}ms
                </div>
                <div className="tooltipCallNodeImplementationTiming">
                  {selfTimeValue === 0
                    ? 'â€”'
                    : `${formatNumberDependingOnInterval(
                        isIntegerInterval,
                        selfTimeValue
                      )}ms`}
                </div>
              </React.Fragment>
            );
          }
        )}
      </div>
    );
  }

  render() {
    const {
      callNodeIndex,
      thread,
      durationText,
      categories,
      callTree,
      timings,
      callNodeInfo: { callNodeTable },
    } = this.props;
    const categoryIndex = callNodeTable.category[callNodeIndex];
    const category = categories[categoryIndex];
    const funcIndex = callNodeTable.func[callNodeIndex];
    const funcStringIndex = thread.funcTable.name[funcIndex];
    const funcName = thread.stringTable.getString(funcStringIndex);

    let displayData;
    if (callTree) {
      displayData = callTree.getDisplayData(callNodeIndex);
    }

    let resourceOrFileName = null;
    // Only JavaScript functions have a filename.
    const fileNameIndex = thread.funcTable.fileName[funcIndex];
    if (fileNameIndex !== null) {
      // Because of our use of Grid Layout, all our elements need to be direct
      // children of the grid parent. That's why we use arrays here, to add
      // the elements as direct children.
      resourceOrFileName = [
        <div className="tooltipLabel" key="file">
          File:
        </div>,
        thread.stringTable.getString(fileNameIndex),
      ];
    } else {
      const resourceIndex = thread.funcTable.resource[funcIndex];
      if (resourceIndex !== -1) {
        const resourceNameIndex = thread.resourceTable.name[resourceIndex];
        if (resourceNameIndex !== -1) {
          // Because of our use of Grid Layout, all our elements need to be direct
          // children of the grid parent. That's why we use arrays here, to add
          // the elements as direct children.
          resourceOrFileName = [
            <div className="tooltipLabel" key="resource">
              Resource:
            </div>,
            thread.stringTable.getString(resourceNameIndex),
          ];
        }
      }
    }

    const stackType = getStackType(thread, funcIndex);
    let stackTypeLabel;
    switch (stackType) {
      case 'native':
        stackTypeLabel = 'Native';
        break;
      case 'js':
        stackTypeLabel = 'JavaScript';
        break;
      case 'unsymbolicated':
        stackTypeLabel = thread.funcTable.isJS[funcIndex]
          ? 'Unsymbolicated native'
          : 'Unsymbolicated or generated JIT instructions';
        break;
      default:
        throw new Error(`Unknown stack type case "${stackType}".`);
    }

    return (
      <div
        className="tooltipCallNode"
        style={{
          '--graph-width': GRAPH_WIDTH + 'px',
          '--graph-height': GRAPH_HEIGHT + 'px',
        }}
      >
        <div className="tooltipOneLine tooltipHeader">
          <div className="tooltipTiming">{durationText}</div>
          <div className="tooltipTitle">{funcName}</div>
          <div className="tooltipIcon">
            {displayData && displayData.icon ? (
              <NodeIcon displayData={displayData} />
            ) : null}
          </div>
        </div>
        <div className="tooltipCallNodeDetails">
          {this._renderTimings(timings, displayData)}
          <div className="tooltipDetails tooltipCallNodeDetailsLeft">
            {/* Everything in this div needs to come in pairs of two in order to
                respect the CSS grid. */}
            <div className="tooltipLabel">Stack Type:</div>
            <div>{stackTypeLabel}</div>
            {/* --------------------------------------------------------------- */}
            <div className="tooltipLabel">Category:</div>
            <div>
              <span
                className={`category-swatch category-color-${category.color}`}
              />
              {category.name}
            </div>
            {/* --------------------------------------------------------------- */}
            {resourceOrFileName}
          </div>
        </div>
      </div>
    );
  }
}
