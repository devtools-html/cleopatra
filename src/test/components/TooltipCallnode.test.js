/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import React from 'react';
import { Provider } from 'react-redux';
import { TooltipCallNode } from '../../components/tooltip/CallNode';
import { render } from 'react-testing-library';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithNativeAllocations } from '../fixtures/profiles/processed-profile';
import { selectedThreadSelectors } from '../../selectors/per-thread';
import * as ProfileSelectors from '../../selectors/profile';
import * as UrlState from '../../selectors/url-state';
import {
  changeSelectedCallNode,
  changeCallTreeSummaryStrategy,
} from '../../actions/profile-view';
import { ensureExists } from '../../utils/flow';

describe('TooltipCallNode', function() {
  function setup(profile) {
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;

    function renderTooltip(): * {
      // This component is not currently connected.
      return render(
        <Provider store={store}>
          <TooltipCallNode
            thread={selectedThreadSelectors.getThread(getState())}
            callNodeIndex={ensureExists(
              selectedThreadSelectors.getSelectedCallNodeIndex(getState()),
              'Unable to find a selected call node index.'
            )}
            callNodeInfo={selectedThreadSelectors.getCallNodeInfo(getState())}
            categories={ProfileSelectors.getCategories(getState())}
            interval={ProfileSelectors.getProfileInterval(getState())}
            durationText="Fake Duration Text"
            callTree={selectedThreadSelectors.getCallTree(getState())}
            callTreeSummaryStrategy={UrlState.getCallTreeSummaryStrategy(
              getState()
            )}
          />
        </Provider>
      );
    }
    return { getState, dispatch, renderTooltip };
  }

  it('handles native allocations', () => {
    const { profile, funcNamesDict } = getProfileWithNativeAllocations();
    const threadIndex = 0;
    const callNodePath = ['A', 'B', 'Fjs', 'Gjs'].map(
      name => funcNamesDict[name]
    );

    const { dispatch, renderTooltip } = setup(profile);
    dispatch(changeSelectedCallNode(threadIndex, callNodePath));
    dispatch(changeCallTreeSummaryStrategy('native-allocations'));
    const { container } = renderTooltip();

    expect(container.firstChild).toMatchSnapshot();
  });
});
