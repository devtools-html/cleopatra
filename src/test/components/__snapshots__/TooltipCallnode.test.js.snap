// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipCallNode handles native allocations 1`] = `
<div
  class="tooltipCallNode"
>
  <div
    class="tooltipOneLine tooltipHeader"
  >
    <div
      class="tooltipTiming"
    >
      Fake Duration Text
    </div>
    <div
      class="tooltipTitle"
    >
      Gjs
    </div>
    <div
      class="tooltipIcon"
    />
  </div>
  <div
    class="tooltipCallNodeDetails"
  >
    <div
      class="tooltipDetails tooltipCallNodeDetailsLeft"
    >
      <div
        class="tooltipLabel"
      >
        Total Bytes:
      </div>
      <div>
        12 bytes
      </div>
      <div
        class="tooltipLabel"
      >
        Self Bytes:
      </div>
      <div>
        5 bytes
      </div>
    </div>
    <div
      class="tooltipDetails tooltipCallNodeDetailsLeft"
    >
      <div
        class="tooltipLabel"
      >
        Stack Type:
      </div>
      <div>
        JavaScript
      </div>
      <div
        class="tooltipLabel"
      >
        Category:
      </div>
      <div>
        <span
          class="colored-square category-color-yellow"
        />
        JavaScript
      </div>
    </div>
  </div>
</div>
`;
